lib/SIL/IR/SILFunctionType.cpp

```diff
@@ -4109,6 +4109,19 @@ static CanSILFunctionType getUncachedSILFunctionTypeForConstant(
     }
   }
 
+  // The type of the native-to-foreign thunk for a swift closure.
+  if (shouldStoreClangType(TC.getDeclRefRepresentation(constant)) &&
+      constant.isForeign && constant.hasClosureExpr()) {
```

Comment:
Could you check the foreign + closure conditions first?  They seem like they'll weed things out a lot more than the shouldStoreClangType condition, which I guess is just future-proofing against other kinds of foreign reference.